### PR TITLE
Do not pin down all dependency versions

### PR DESCRIPTION
**Pull request by silviot**
This PR implements the proposal documented in #2610.

Two tests still fail in CI.

They pass on my machine if I invoke them with
```
pytest -rsfx tests/test_experiment_server.py -k test_failure  --runslow 
pytest -k test_fails_if_run_outside_experiment_dir -rsfx --runslow
```

@jessesnyder Do you have a hunch about what could cause these failures?

### CODE REVIEW

**Comment by silviot** [comment]
This PR is based on #2615, and is meant to be reviewed and merged after it.

**Comment by jessesnyder** [comment]
@silviot The MTurk one ("tests.test_mturk.FixtureConfigurationError: No "mturk_worker_id" value found. Either set this value or skip these tests with `pytest -m "not mturkworker"`") is a result of the fact that we don't have an MTurk Sandbox worker set up for the test to use. @jacobyn, if that's the case, is it still on your radar?

The other one downloads a known file from a known S3 bucket, which must not be accessible from the Github action context for some reason.

**Comment by jacobyn** [comment]
I am happy to set further Mturk credentials. What exactly we need?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ repos:
   hooks:
     - id: black
       language_version: python3 # Should be a command that runs python3.6+
-      files: ^(tests|dallinger|dallinger_scripts|demos)/
+      files: ^(tests|dallinger|dallinger_scripts|demos)/|setup.py
```
Comment (jessesnyder) [inline .pre-commit-config.yaml:8]: 👍

File: dallinger/deployment.py [hunk 2]
```diff
@@ -76,7 +77,7 @@ def files(self):
         """A Set of all files copyable in the source directory, accounting for
         exclusions.
         """
-        return {path for path in self._walk()}
+        return set(self._walk())
```
Comment (jessesnyder) [inline dallinger/deployment.py:80]: I confess I think in comprehensions :)

File: tests/test_deployment.py [hunk 3]
```diff
@@ -845,7 +852,7 @@ def test_failure(self, debugger):
 @pytest.mark.slow
 @pytest.mark.docker
 class TestDockerServer(object):
-    @pytest.mark.usefixtures("check_runbot")
+    @pytest.mark.skipif(bool(os.environ.get("CI")), reason="Fails when run in the CI")
```
Comment (jessesnyder) [inline tests/test_deployment.py:855]: 👍

**Review by jessesnyder — COMMENTED** [review]
Looks good so far! I'll look more closely next week.

**Comment by jessesnyder** [comment]
@silviot I wasn't able to run the Bartlett demo on Heroku using this branch. Looks like the entry-point scripts weren't installed maybe?
```
2021-03-16T03:46:12.983425+00:00 app[api]: Scaled to console@0:Hobby web@1:Hobby worker@1:Hobby by user jesse@rasikaconsulting.com
2021-03-16T03:46:21.442859+00:00 heroku[web.1]: Starting process with command `dallinger_heroku_web`
2021-03-16T03:46:24.518285+00:00 app[web.1]: bash: dallinger_heroku_web: command not found
2021-03-16T03:46:24.587314+00:00 heroku[web.1]: Process exited with status 127
2021-03-16T03:46:30.072624+00:00 heroku[web.1]: Starting process with command `dallinger_heroku_web`
2021-03-16T03:46:32.900967+00:00 heroku[web.1]: Starting process with command `dallinger_heroku_web`
2021-03-16T03:46:33.727601+00:00 app[web.1]: bash: dallinger_heroku_web: command not found
2021-03-16T03:46:33.810245+00:00 heroku[web.1]: Process exited with status 127
2021-03-16T03:46:35.860618+00:00 app[web.1]: bash: dallinger_heroku_web: command not found
2021-03-16T03:46:35.935149+00:00 heroku[web.1]: Process exited with status 127
2021-03-16T03:46:36.012104+00:00 heroku[web.1]: State changed from starting to crashed
2021-03-16T03:46:36.019032+00:00 heroku[web.1]: State changed from crashed to starting
2021-03-16T03:46:44.189141+00:00 heroku[worker.1]: Starting process with command `dallinger_heroku_worker`
2021-03-16T03:46:44.916694+00:00 heroku[worker.1]: State changed from starting to up
2021-03-16T03:46:47.440317+00:00 app[worker.1]: bash: dallinger_heroku_worker: command not found
2021-03-16T03:46:47.507627+00:00 heroku[worker.1]: Process exited with status 127
2021-03-16T03:46:47.578631+00:00 heroku[worker.1]: State changed from up to crashed
2021-03-16T03:46:47.582472+00:00 heroku[worker.1]: State changed from crashed to starting
2021-03-16T03:46:54.003001+00:00 heroku[web.1]: Starting process with command `dallinger_heroku_web`
2021-03-16T03:46:56.770530+00:00 app[web.1]: bash: dallinger_heroku_web: command not found
2021-03-16T03:46:56.826383+00:00 heroku[web.1]: Process exited with status 127
2021-03-16T03:46:56.905629+00:00 heroku[web.1]: State changed from starting to crashed
2021-03-16T03:46:59.738734+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=POST path="/launch" host=dlgr-9270d4d2.herokuapp.com request_id=b134cae3-f506-4bed-8012-de3d016d3a55 fwd="97.113.208.0" dyno= connect= service= status=503 bytes= protocol=https
2021-03-16T03:47:00.003435+00:00 heroku[worker.1]: Starting process with command `dallinger_heroku_worker`
2021-03-16T03:47:00.747153+00:00 heroku[worker.1]: State changed from starting to up
2021-03-16T03:47:01.988145+00:00 app[worker.1]: bash: dallinger_heroku_worker: command not found
2021-03-16T03:47:02.041539+00:00 heroku[worker.1]: Process exited with status 127
2021-03-16T03:47:02.107871+00:00 heroku[worker.1]: State changed from up to crashed
```

**Comment by silviot** [comment]
@jessesnyder That error was caused by wrong `constraints.txt` files in the experiment directory.

Now they're fixed, and there's an additional `dallinger generate-constraints` command to (re)generate them.

Also, this PR now packages the current dallinger wheel (only if it was installed as editable with `-e`) and uploads it to Heroku.

**Review by jessesnyder — APPROVED** [review]
Deployed memoryexp3 successfully on Heroku, and tests are passing, so I think we can merge this. We will need to update documentation around [re]generating constraints files, but I think this can be merged ahead of that to help us make progress in other areas.

**Pull request merged**